spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubInboundChannelAdapter.java

```diff
@@ -51,17 +49,13 @@
 
 	private AckMode ackMode = AckMode.AUTO;
 
-	private MessageConverter messageConverter;
+	private Converter<byte[], Object> payloadConverter;
```

Comment:
I think this change is not true.
We still have to apply here a `MessageConverter.toMessage(Object payload, @Nullable MessageHeaders headers)`.
The difference with Kinesis that we don't have there headers to consult what target payload type to choose. That's why a plain `Converter<byte[], ?>` only can be applied there.

Here it is with a slightly different story, where we really map headers from the `PubSubMessage`.

Let's consider to comeback to the `MessageConverter`. Not sure if that still should be a mentioned `StringMessageConverter` or something different, but we definitely can restore an old behavior and reconsider an *inbound* part in the separate issue.

For the *outbound* side we strictly know that the target type is a `byte[]` and it doesn't matter for us what is the payload type - the provide `Converter<Object, byte[]>` should take care about that. That's why I don't see reason to manage the logic there with the `MessageConverter`. 

If you want, let's come back to the `MessageConverter` there as well for symmetry! Only the problem that we need to introduce some default implementation with the serialization capabilities...

Reply:
@artembilan I considered the header conversion issue as well, but since we already have a `headerMapper` in this class, I think it's really OK, and even better to just use the payload converter. WDYT?

Reply:
No, you didn't get it. We receive a raw `byte[]` and headers may have an information into what type to covert it for the message to send. That's exactly what a `MessageConverter.toMessage(Object payload, @Nullable MessageHeaders headers)` suggests for us. With the raw `Converter` we don't have such an ability and maximum what can be useful for us is a standard Java Serialization model.

So, IMHO that's not OK in this case and we really should come back to the `MessageConverter` on the inbound side.

Reply:
I see your point, but here's my concern. If we have `MessageConverter`, then what do we do with the `headerMapper`, because it seems confusing to use a `MessageConverter` just to convert the payload, but not the whole message. The other nuisance is the asymmetry between the inbound and the outbound. Can we just say that conversion dependent on headers is an advanced use-case that should be handled outside of the adapter somewhere else in the integration flow?
Otherwise, another way to say what you're asking for is `Converter<Message, byte[]>`, which I guess we can do without switching to `MessageConverter`.

Reply:
Not sure how a `HeaderMapper` confuses you since it does only one purpose: creates a plain Java `Map` from the native header representation: `Map<String, Object> toHeaders(T source)`. As you see there is nothing about a `Message` at all. The `MessageConverter` is really the next step in the logic to build a proper `Message` from the native payload and converted headers.

The `Converter<Message, byte[]>` still doesn't bring too much value since you don't provide `headers` contract and don't propagate them to the converter.

Ass I said: we can bring `MessageConverter` to the outbound side as well to let you do not be confused, but that doesn't bring any values since we always expect only a `byte[]` as outcome from the conversion.

I don't see reason to document anything as a limitation since we have handy `HeaderMapper` and `MessageConverter` as out-of-the-box solution in Spring already.

Please, take a look to existing Channel Adapter implementations. I would say we have already came through similar architectural concerns in other places and that's how we ended up with the 
`HeaderMapper` and `MessageConverter`  pair. The outbound side really can be questionable, but IMO it's overheader to bring `MessageConverter` there since we always expect only `byte[]`.

Reply:
I was actually looking at the outbound, while you were talking about the inbound. Sorry!
Now what you are saying makes sense! :blush: